cpp/unreal_plugins/UrdfRobot/Source/UrdfRobot/UrdfJointComponent.cpp

```diff
@@ -0,0 +1,487 @@
+    {"One",        {"add_torque_in_radians",             { 1000.0,  0.0,  0.0}}},
+    {"Two",        {"add_torque_in_radians",             {-1000.0,  0.0,  0.0}}},
+    {"Three",      {"add_force",                         {   50.0,  0.0,  0.0}}},
+    {"Four",       {"add_force",                         {  -50.0,  0.0,  0.0}}},
+    {"Five",       {"add_to_angular_velocity_target",    {    0.1,  0.0,  0.0}}},
+    {"Six",        {"add_to_angular_velocity_target",    {   -0.1,  0.0,  0.0}}},
+    {"Seven",      {"add_to_linear_velocity_target",     {    0.1,  0.0,  0.0}}},
+    {"Eight",      {"add_to_linear_velocity_target",     {   -0.1,  0.0,  0.0}}},
+    {"Nine",       {"add_to_angular_orientation_target", {    0.0,  0.0,  2.0}}},
+    {"Zero",       {"add_to_angular_orientation_target", {    0.0,  0.0, -2.0}}},
+    {"Underscore", {"add_to_linear_position_target",     {    2.0,  0.0,  0.0}}},
+    {"Equals",     {"add_to_linear_position_target",     {   -2.0,  0.0,  0.0}}}
+};
+
+UUrdfJointComponent::UUrdfJointComponent()
+{
+    SP_LOG_CURRENT_FUNCTION();
+
+    // UInputActionComponent
+    input_action_component_ = CreateDefaultSubobject<UInputActionComponent>(Unreal::toFName("input_action_component"));
+    SP_ASSERT(input_action_component_);
+    // Need to explicitly set this up so that the component hierarchy is well-defined.
+    input_action_component_->SetupAttachment(this);
+}
+
+UUrdfJointComponent::~UUrdfJointComponent()
+{
+    SP_LOG_CURRENT_FUNCTION();
+
+    JointType = EJointType::Invalid;
+    JointControlType = EJointControlType::NotActuated;
+    EnableKeyboardControl = false;
+    //LinearTranslationOffset = FVector::ZeroVector; // TODO (MR): support linear translation offsets
+
+    parent_static_mesh_component_ = nullptr;
+    child_static_mesh_component_ = nullptr;
+
+    SP_ASSERT(input_action_component_);
+    input_action_component_ = nullptr;
+}
+
+void UUrdfJointComponent::BeginPlay()
+{
+    UPhysicsConstraintComponent::BeginPlay();
+
+    const std::map<std::string, std::pair<std::string, std::vector<double>>> input_actions = INPUT_ACTIONS;
+    input_action_component_->bindInputActions(input_actions);
+    input_action_component_->apply_input_action_func_ = [this, input_actions](const std::string& key) -> void {
+        if (EnableKeyboardControl) {
+            std::pair<std::string, std::vector<double>> action_component = input_actions.at(key);
+            // only assert if we're not in the editor
+            bool assert_if_action_is_inconsistent_with_joint = !WITH_EDITOR;
+            applyAction(action_component.first, action_component.second, assert_if_action_is_inconsistent_with_joint);
+        }
+    };
+}
+
+std::map<std::string, ArrayDesc> UUrdfJointComponent::getActionSpace() const
+{
+    std::map<std::string, ArrayDesc> action_space;
+
+    if (JointType == EJointType::Invalid || JointType == EJointType::Fixed) {
```

Comment:
- The logic in this if statement is incorrect. It should assert if `Invalid` for consistency with the rest of this class, which asserts in the case of an invalid joint.
- Roll this separate if-statement into the rest of the switch statement for consistency with the rest of this class.

Reply:
Updated the `Invalid` case as suggested.

For `EJointType::Fixed` type, the reason I'm returning an empty `action_space` is that if we do not return an empty map, the description with a `{<name_corresponding_to_fixed_joint>: ArrayDesc}` will be forwarded to python client as part of the `action_space`. And, the python client has to send this action as an entry in the action dict in every step back to the UE, even though this is a `fixed` joint and it cannot be moved. Hence, I'm only exposing those action's that can be acted upon, unlike `fixed`.

This does bring in a layer of additional filtering, which might not be desirable. What do you think is better here?
1. Exposing action space of `fixed` joints even though they are not useful from the perspective of applying actions. But, makes the code a bit simpler - no filtering, just iterate over all joint types and create a description.
2. Don't expose `fixed` joints, as in the current code.

Reply:
I agree that a fixed joint should return `{}` as its action space. Here is what I'm suggesting.

```
    std::map<std::string, ArrayDesc> action_space;

    ArrayDesc array_desc;
    array_desc.low_ = std::numeric_limits<double>::lowest();
    array_desc.high_ = std::numeric_limits<double>::max();
    array_desc.shape_ = {3};
    array_desc.datatype_ = DataType::Float64;

    switch (JointControlType) {
        case EJointControlType::Position:
        case EJointControlType::PositionAndVelocity:
            switch (JointType) {
                case EJointType::Continuous:
                case EJointType::Revolute:
                    action_space[Unreal::toStdString(GetName()) + "." + "whatever"] = std::move(action_space);
                    break;

                case EJointType::Prismatic:
                    action_space[Unreal::toStdString(GetName()) + "." + "whatever"] = std::move(action_space);
                    break;

                case EJointType::Fixed:
                    break;

                default:
                    SP_ASSERT(false); // TODO (MR): support planar joints
                    break;
            }
            break;

        // ...
    }

    return action_space;
}
```